### ISSUE TITLE
Skyrim/Oblivion Updates

### ISSUE THREAD

**Issue by Sharlikran** [Sharlikran/wrye-bash#25]
https://github.com/Sharlikran/wrye-bash/tree/dev-skyrim-patcher-update contains all my Skyrim/Oblivion changes.  It is rebased off of dev.

Please let me know if you would like me to rebase or change any code.

I only ask it keeps the same functionality, the changes are dumbed-down to Python programming level 0 so I can understand them, and that I don't have to read python syntax documentation.  Aside from that I am happy to make any changes you like.

**Comment by Sharlikran** [comment]
@Utumno these are my Skyrim/Oblivion changes so far

**Comment by Utumno** [comment]
Thanks ! I will look into those more carefully in the next week (?) - for now don't spend energy revising saving reading code as it needs a full rewrite anyway (actually I started doing this but it's tricky and it  will take some time plus not a priority for the beta - just with the work on bsas I have some ideas of how it should be done correctly)

**Comment by Sharlikran** [comment]
No rush as the code does what it needs to do for now. I have a request to add utf8 support for save game files and the person seems to know the code used.  I'll add that once its done.

I posted the code to be respectfull and to show I mean it when I say I'm grateful for your work.  It would be disapointing if it were not used in its entirety.  Whatever you decide for wrye- bash is up  to you.  Myself and others appreciate the work you have done it has given Wrye a noticeable performance boost.

I'll keep using my code as is though because all the changes are needed in order to have the best gaming experience.  Which is what my goal has been for the last 5 years.  I provide the best gaming experience for Nexus users and put their interests first. My only restriction being that I can't in Python do I can't always add all the functionality needed.

Keep up the good work.

**Comment by Sharlikran** [comment]
And to clarify i will be trying to not use the cut and pasted code for constants and records unless it just becomes too confusing.  It will take a few days to get that all removed though.

**Comment by Utumno** [comment]
Ok thanks sharl - just compare the file I do `import *` from and your code to see what differs - hopefully and probably nothing.

If any pythonista is reading this - import * is evil, used here exceptionally

**Comment by Sharlikran** [comment]
Not a problem.  I am going to make every effort to eliminate the cut and past 20,000 lines.  I may need to request not to loose other code but hopefully we can come to some agreement as the code I'm using address user concerns and posts.  You have no time to answer nexus posts and I simply don't want to keep answering the same ones over and over as I really don't have the time either.

**Comment by Utumno** [comment]
Thanks sharl - I mocked somehow skyrim se so I am a bit confident so I will merge to dev so you will have a solid base to compare to.

Don't put too much effort till the dust settles, hopefully end of the week

Your addressing posts is really helpful - you could note in the commit message the post as I do. Plus one commit for any such issue would be then easy to review and merge to dev.

Please avoid changing whitespace, order of constants etc - it is really adding a lot of noise in the reviews to the point of impossible - especially cause it's frustrating to realize hey this 200 lines changes are mostly moved around code

**Comment by Sharlikran** [comment]
I understand it's best not to change the order of the code.  When I compared all the files between the different games originally I made sure a long time ago that they had the exact same order.  I didn't move them from the order I set them to.  I'll check the order is the way I originally had it and as I update the records or add new record definitions and constants (that are actually used and not just cut and pasted) I'll continue to make sure I don't change the order I set things to so you don't get all that commit noise and so you can easily compare differences as well.  Thanks for the reminder.

**Comment by Utumno** [comment]
Yep thanks - I know that you know - mentioning cause this was a problem in the fallout4 support branch here in the constants.py - and that's a priority cause I do want all the missing ones on dev

**Comment by Sharlikran** [comment]
I think the main problem with the Fallout4 files is there were multiple contributors.  At least I think so.  People saying this should be in this order and this in this order, when that's not the order I put it in.  I'm not trying to tell you that it's your fault, accuse you, or order you around, but those long lists have to be 100% accurate and the list should be made by me since I have the game.  Then I can verify them, run the Clean Data procedure to test it.  Then when you want the strings files removed, you would simply ask me to remove them instead of doing it yourself, and then order won't change and it won't mess up your commits with all that noise.

I know I don't always change some code the way you want.  However, I will be making an attempt to do 80% of what you ask.  I just ask some leeway and decision making when it comes to some things please.  Especially with all the game files for the different games.  And for the addition of the messages when Bash initializes for Skyrim SE and the warnings window.  Just simple stuff like that makes a big difference.  

And again it's definitely not a problem to eliminate the cut and pasted code even though it isn't hurting anything being there.

**Comment by Utumno** [comment]
> I think the main problem with the Fallout4 files is there were multiple contributors

Yes but the general rule is the code as is in dev is the common base - so if for one reason or the other this code ended up with that order then contributors should rebase off that.

>  Then when you want the strings files removed, 

If utumno-wip with current ba2 code detects the string files alright we should remove them as the latest patch of fallout 4 does not have those. Btw a task - add to the readme here (and meta I guess) info on which patch of the games we support!

>  And for the addition of the messages when Bash initializes for Skyrim SE and the warnings window. Just simple stuff like that makes a big difference

Will have a look - keep in mind that this breaks existing translations and that's one (of the) reason(s) I avoid changing strings. But yes since you are the one interacting with users I will certainly consider changing those strings :)

> even though it isn't hurting anything being there.

Nope it does - but that's the technical side so let this up to me ;)

**Comment by Sharlikran** [comment]
> Will have a look - keep in mind that this breaks existing translations and that's one (of the) reason(s) I avoid changing strings. But yes since you are the one interacting with users I will certainly consider changing those strings :)

That does make sense but at this point I highly doubt any of the translations have been updated for years.  In addition to that if I translated those from "corrupted" to Spanish I would have to choose a different word anyway because corrupted in Spanish is more like when the police are corrupt and working with criminals and receiving some kind of money to look the other way for certain criminal activity.

I'm not trying to be nitpicky or argue but even valda was going to make a Japanese translation and stopped.  I highly doubt any of that is up to date.  Since I deal with users that see the messages in English they are concerned with the strong messages.  One user got into an argument with me while others were simply frighted that the message was there and wondering if the mods were safe to use or not.

Right now with Skyrim SE the hot questions is "is it safe" to do this or use this plugin.  Nothing has changed for cleaning files but it's aswked at least once a day "is it safe to clean files".  People even ask does the old video still apply.  Most users feel if it's for Skyrim then it's outdated and doesn't apply and look for an SSE versions.  It's just the Nexus mentality.  Because the CK is different, the plugins have had changes, NIF files don't work, and a lot of other stuff everyone wants to know is "is it safe" and seeing a message "this is corrupt" puts them in a state of panic.  They trust the tools they have used in the past to be accurate.  Since the old Skyrim (306 Wrye Bash) never said that they feel with Skyrim SE the message must mean something is wrong.  When they see the strings files are missing the also panic because they never had to extract the strings before, so why do they have to do it now?  Furthermore will that affect their game and how?

Nexus users come in all shapes and skill levels.  What makes sense to you scares others.  What seems easy to you is the hardest thing they have ever done and they need a video to do it correctly.  I even had to make a video on how to extract the strings files because some could not follow these simple steps.

1. Make a temp folder for the files.
2. Run BSAManager.exe and click the small blue folder to open a file
3. Navagate to your Skyrim SE Data folder and choose Skyrim - Patch.bsa
4. Select the files you want and right click them
5. Choose the temp folder as the destination to extract the files
6. Be sure to get the Skyrim and DLC strings files
7. You will need to make a Strings folder in your Skyrim SE Data folder
8. Copy the strings files to .\Data\Strings for the language you need i.e. English.

Nexus users just don't get it sometimes and you have to cater to the lowest common denominator but in a respectful way.

**Comment by Utumno** [comment]
Ok thanks - will have a look asap - let me finalize bsas reading first so we don't have to extract the strings - my branch does it for se I think but fallout 4 ?

Also adding a commit editing the readme to mention fallout se support and the number of patch we support for each game would help - I don't play the games so info like this escapes me...

**Comment by Sharlikran** [comment]
@Utumno 

- Removed INI Tweak Files
- Updated NSI Info - Should not remove folder or user created files

**Comment by Utumno** [comment]
Thanks - the other stuff I merged includes some stuff in this branch so at your convenience rebase (namely the tweaks removal is already on dev). Also please edit the readmes (if needed) to remove the mentions to Mopy/Ini Tweaks

**Comment by Utumno** [comment]
Ok will include your commit on warnings but needs edits - corrupted mods are not only versions mismatches (see ModInfos.refresh() and parent method) and I don't know why :

```

             self.knownCorrupted |= corruptSaves
-        invalidVersions = set([x.name for x in bosh.modInfos.values() if round(
-            x.header.version, 6) not in bush.game.esp.validHeaderVersions])
+        invalidVersions = set([x.name for x in bosh.modInfos.values() if float(
+            "{0:.2f}".format(x.header.version)) not in bush.game.esp.validHeaderVersions])
```

when you make a change like that please give info on the commit message - the same round() call is used in other places so should be everywhere if you found it breaks stuff  - does it ?

**Comment by Sharlikran** [comment]
> when you make a change like that please give info on the commit message

Done. I added a message to commit https://github.com/Sharlikran/wrye-bash/commit/a482ac285b0d5ca1726b2c5adbb15ed7467118ed.

> the same round() call is used in other places so should be everywhere if you found it breaks stuff  - does it ?

I can't use PyCharm nor do I know python nor would I know if other places are broken.  This is to ensure that the number returned is x <= y, because sometimes the comparison is 1.70 <= 1.70000393

If python is only using the 2nd decimal place when using round() that's fine but I didn't see that when I use print(str(invalidVersions ))

I understand your concerns but this has been working since Oct 27 and has been used by approximately 11,729 people.  I'm pretty sure it's solid.

**Comment by Utumno** [comment]
> I can't use PyCharm nor do I know python nor would I know if other places are broken

Please try to set it up - tehre are instructions in the wiki - after the initial leaarning curve you'll find it really helps

**Comment by Sharlikran** [comment]
I may not do that as it isn't part of my recent changes in how I approach the project.  Which we have discussed.  It would fall within Programing and Development.  Which as I have stated I do not intend to do.  I provide these changes because they are beneficial to myself, the Wrye Project, and the many Wrye enthusiasts using Wrye Bash.

**Comment by Sharlikran** [comment]
Although I will add this.  You push me to improve my Git knowledge which has come in handy.  I am sure it is now paying off as I would bet my commits are somewhat easier (maybe not perfect) to use. and probably cause you less headaches.  Because of your efforts I have learned much and I do appreciate it.

I may try to improve with my use of Pycharm, but I make no promises.  That's not even a safe statement so I don't have to improve.  Development on the same level as you in regards to Wrye is simply not something I wish to do.  Since I know it does make things easier for you, which is something I care about, we will see.

**Comment by Sharlikran** [comment]
I even use gitk much more then I used to, maybe not how you would like, but much more then I used to.

**Comment by Utumno** [comment]
I am sure you do Sharl.
